src/component/Flow.tsx

```diff
@@ -1,366 +1,131 @@
-import * as FlipMove from 'react-flip-move';
-import update from 'immutability-helper';
-import { v4 as generateUUID } from 'uuid';
-import { FlowDefinition, Action, Position, SendMsg, Node, UINode, Dimensions } from '../flowTypes';
-import ComponentMap from '../services/ComponentMap';
-import NodeComp, { DragPoint } from './Node';
-import FlowMutator from '../services/FlowMutator';
-import SimulatorComp from './Simulator';
-import Plumber from '../services/Plumber';
-import ActivityManager from '../services/ActivityManager';
-import NodeEditor, { NodeEditorProps } from './NodeEditor';
-import LanguageSelectorComp, { Language } from './LanguageSelector';
-import { ConfigProviderContext, endpointsPT } from '../config';
+import { react as bindCallbacks } from 'auto-bind';
+import { connect } from 'react-redux';
+import { Config } from '../config';
 import { getActivity } from '../external';
-import { snapToGrid } from '../utils';
-
+import { FlowDefinition, Languages, Node, Position, UINode, AnyAction } from '../flowTypes';
+import {
+    Components,
+    ConnectionEvent,
+    Constants,
+    DispatchWithState,
+    ensureStartNode,
+    onConnectionDrag,
+    onOpenNodeEditor,
+    ReduxState,
+    resetNewConnectionState,
+    updateConnection,
+    updateCreateNodePosition
+} from '../redux';
+import ActivityManager from '../services/ActivityManager';
+import Plumber from '../services/Plumber';
+import { getConnectionError, snapToGrid } from '../utils';
 import * as styles from './Flow.scss';
+import NodeContainer, { DragPoint } from './Node';
+import NodeEditor from './NodeEditor';
 
 export interface FlowProps {
-    nodeDragging: boolean;
-    onDrag(dragging: boolean): void;
-    language: Language;
     translating: boolean;
     definition: FlowDefinition;
     dependencies: FlowDefinition[];
-    Mutator: FlowMutator;
-    ComponentMap: ComponentMap;
-}
-
-export interface FlowState {
-    ghost?: Node;
-    nodeEditor?: NodeEditorProps;
-    loading: boolean;
-    viewDefinition?: FlowDefinition;
-}
-
-export interface Connection {
-    previousConnection: Connection;
-}
-
-export interface ConnectionEvent {
-    connection: Connection;
-    source: Element;
-    target: Element;
-    sourceId: string;
-    targetId: string;
-    endpoints: any[];
+    languages: Languages;
+    nodes: Node[];
+    components: Components;
+    ghostNode: Node;
+    pendingConnection: DragPoint;
+    nodeEditorOpen: boolean;
+    ensureStartNodeAC: () => void;
+    updateConnectionAC: (source: string, target: string) => void;
+    onOpenNodeEditorAC: (node: Node, action: AnyAction, languages: Languages) => void;
+    resetNewConnectionStateAC: () => void;
+    onConnectionDragAC: (event: ConnectionEvent) => void;
+    updateCreateNodePositionA: (
+        createNodePosition: Position
+    ) => { type: Constants; payload: { createNodePosition: Position } };
 }
 
 export interface Translations {
     [uuid: string]: any;
 }
 
-const REPAINT_DURATION = 600;
```

Comment:
this is the application of the `Config` component I was talking about, it's actually pretty similar to the `Consumer` component React's new context API provides. none of those ugly errors for components that consume context in this way